## Imports

In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt
import os
import sys


In [2]:
data_path = 'data/WA_Fn-UseC_-Telco-Customer-Churn.csv'

In [3]:
df = pd.read_csv(data_path)
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
customerID          7043 non-null object
gender              7043 non-null object
SeniorCitizen       7043 non-null int64
Partner             7043 non-null object
Dependents          7043 non-null object
tenure              7043 non-null int64
PhoneService        7043 non-null object
MultipleLines       7043 non-null object
InternetService     7043 non-null object
OnlineSecurity      7043 non-null object
OnlineBackup        7043 non-null object
DeviceProtection    7043 non-null object
TechSupport         7043 non-null object
StreamingTV         7043 non-null object
StreamingMovies     7043 non-null object
Contract            7043 non-null object
PaperlessBilling    7043 non-null object
PaymentMethod       7043 non-null object
MonthlyCharges      7043 non-null float64
TotalCharges        7043 non-null object
Churn               7043 non-null object
dtypes: float64(1), int64(2), obj

In [6]:
df.Churn.value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

In [7]:
df.Churn.value_counts(normalize=True)

No     0.73463
Yes    0.26537
Name: Churn, dtype: float64

In [45]:
#new helper method:
def Counts(str):
    print(str)
    return df.str.value_counts()

In [8]:
print(df.MultipleLines.value_counts())
print(df.InternetService.value_counts())
print(df.PaymentMethod.value_counts())
print(df.OnlineBackup.value_counts())
print(df.DeviceProtection.value_counts())

No                  3390
Yes                 2971
No phone service     682
Name: MultipleLines, dtype: int64
Fiber optic    3096
DSL            2421
No             1526
Name: InternetService, dtype: int64
Electronic check             2365
Mailed check                 1612
Bank transfer (automatic)    1544
Credit card (automatic)      1522
Name: PaymentMethod, dtype: int64
No                     3088
Yes                    2429
No internet service    1526
Name: OnlineBackup, dtype: int64
No                     3095
Yes                    2422
No internet service    1526
Name: DeviceProtection, dtype: int64


Perhaps its worth investigating those 1526 people in the future

## Data Cleaning

I need numbers, not strings

In [59]:
df['gender'].replace('Male', 0, inplace = True)
df['gender'].replace('Female', 1, inplace = True)

male = 0, female = 1

In [9]:
def strToInt(column):
    for i in range(len(df)):
        if df[column][i] == 'Yes':
            df[column][i] = 1
        elif df[column][i] == 'No':
            df[column][i] = 0
        else:
            df[column][i] = 2

def makeBool(column):
    df[column].replace(('Yes', 'No'), (1, 0), inplace = True)

def prepDF():
    df['gender'].replace(('Male', 'Female'), (0, 1), inplace = True)
    makeBool('Partner')
    makeBool('Dependents')
    makeBool('PhoneService')
    makeBool('PaperlessBilling')
    makeBool('Churn')
    strToInt('MultipleLines')
    strToInt('OnlineSecurity')
    strToInt('OnlineBackup')
    strToInt('DeviceProtection')
    strToInt('TechSupport')
    strToInt('StreamingTV')
    strToInt('StreamingMovies')
    makeBool('InternetService')
    df['InternetService'].replace(('DSL', 'Fiber optic'), (1, 2), inplace = True)
    df['Contract'].replace(('Month-to-month', 'One year', 'Two year'), (0, 1, 2), inplace = True)
    df['PaymentMethod'].replace(('Electronic check', 'Mailed check', 'Bank transfer (automatic)', 'Credit card (automatic)'), (0, 1, 2, 2), inplace = True)
    
    

In [10]:
prepDF()
df.head()

/Users/adamroth/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/adamroth/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/adamroth/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after re

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        1           0       1   
1  5575-GNVDE       0              0        0           0      34   
2  3668-QPYBK       0              0        0           0       2   
3  7795-CFOCW       0              0        0           0      45   
4  9237-HQITU       1              0        0           0       2   

   PhoneService MultipleLines  InternetService OnlineSecurity  ...  \
0             0             2                1              0  ...   
1             1             0                1              1  ...   
2             1             0                1              1  ...   
3             0             2                1              1  ...   
4             1             0                2              0  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies Contract  \
0                0           0           0               0        0   
1                1           0           0               0        1   
2                0           0           0               0        0   
3                1           1           0               0        1   
4                0           0           0               0        0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
0                 1              0           29.85         29.85     0  
1                 0              1           56.95        1889.5     0  
2                 1              1           53.85        108.15     1  
3                 0              2           42.30       1840.75     0  
4                 1              0           70.70        151.65     1  

[5 rows x 21 columns]

## Scikit Learn Imports

In [53]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler

from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

In [54]:
le = LabelEncoder()

df['Churned'] = le.fit_transform(df['Churn'])

df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  Churn  Churned  
0    No        0  
1    No        0  
2   Yes        1  
3    No        0  
4   Yes        1

In [55]:
X = df.drop(['customerID', 'Churn', 'Churned'],  axis=1)
y = df['Churned']

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=7)

In [56]:
ss = StandardScaler()

X_test_ss = ss.fit(X_test)
X_train_ss = ss.fit_transform(X_train)


ValueError: could not convert string to float: 'Female'